src/v/utils/vint.h

```diff
@@ -12,38 +12,73 @@
 #pragma once
 #include "bytes/bytes.h"
 
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/iostream.hh>
+
 #include <cstdint>
 
 namespace unsigned_vint {
 /// At most 5 bytes are needed to encode a 32 bit value
 inline constexpr size_t max_length = 5;
 
 namespace detail {
-
-/// \brief almost identical impl to leveldb, made generic for c++
-/// friendliness
-/// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
-template<typename Range>
-inline std::pair<uint64_t, size_t>
-deserialize(Range&& r, uint8_t limit) noexcept {
+struct var_decoder {
     uint64_t result = 0;
     size_t bytes_read = 0;
     uint64_t shift = 0;
-    for (auto src = r.begin(); shift <= limit && src != r.end(); ++src) {
-        uint64_t byte = *src;
+    uint8_t limit;
+
+    explicit var_decoder(uint8_t limit)
+      : limit(limit) {}
+
+    /**
+     * @brief Incorporate a new byte into the state.
+     * Returns true if the outer loop should break (last byte found).
+     */
+    bool accept(uint8_t byte) {
+        /// \brief almost identical impl to leveldb, made generic for c++
+        /// friendliness
+        /// https://github.com/google/leveldb/blob/201f52201f/util/coding.cc#L116
         bytes_read++;
         if (byte & 128) {
             result |= ((byte & 127) << shift);
         } else {
             result |= byte << shift;
-            break;
+            return true;
         }
         shift += 7;
+        return shift > limit;
+    }
+};
+
+template<typename Range>
+inline std::pair<uint64_t, size_t>
+deserialize(Range&& r, uint8_t limit) noexcept {
+    var_decoder decoder(limit);
+    for (auto byte : r) {
+        if (decoder.accept(static_cast<uint8_t>(byte))) {
+            break;
+        }
     }
-    return {result, bytes_read};
+    return std::make_pair(decoder.result, decoder.bytes_read);
 }
+
 } // namespace detail
 
+inline ss::future<std::pair<uint32_t, size_t>>
+deserialize(ss::input_stream<char>& s) {
+    detail::var_decoder decoder(5 * sizeof(uint64_t));
+    while (!s.eof()) {
+        auto buf = co_await s.read_exactly(1);
```

Comment:
As discussed, another area for perf optimization in the future if we need it. My first thought was; if you can get short reads without exceptions here, you can just request a bigger amount and see what you get back. +1 for now (correctness first then perf can be followup.)

Reply:
Linking @travisdowns awesome detailed write up about this from slack [here](https://redpandadata.slack.com/archives/C01ND4SVB6Z/p1651795124339989?thread_ts=1651708685.854219&cid=C01ND4SVB6Z)